### PR TITLE
Add middleware to process multiple tokens concurrently

### PR DESCRIPTION
**Pull request by JAORMX**
This is done through a new middleware structure called
`MultiTokenMiddleware`, which will concurrently validate tokens against
multiple endpoints. These endpoints are meant to be generic, but an
initial implementation is the JWT middleware that we already have in
this repo.

### CODE REVIEW

**Comment by JAORMX** [comment]
This is no longer needed.

**Comment by JAORMX** [comment]
We might need this after all

File: ginauth/genericmiddleware.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package ginauth
+
+import (
+	"github.com/gin-gonic/gin"
+)
+
+// ClaimMetadata returns the minimal relevant information so middleware
+// can set the appropriate metadat to a context (e.g. a gin.Context)
```
Comment (sfunkhouser) [inline ginauth/genericmiddleware.go:8]:
```suggestion
// can set the appropriate metadata to a context (e.g. a gin.Context)
```

File: ginauth/multitokenmiddleware.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+				if err != nil {
+					v.SetMetadata(c, cm)
+				}
+
+				r <- err
+			}(verifier, c, res)
+		}
+
+		wg.Wait()
```
Comment (sfunkhouser) [inline ginauth/multitokenmiddleware.go:67]: Do we need to enforce some timeout so if one doesn't return it doesn't hang?

Comment (JAORMX) [reply ginauth/multitokenmiddleware.go:67]: Yes, this could probably be handled by each Middleware though.

Comment (JAORMX) [reply ginauth/multitokenmiddleware.go:67]: It's now handled by the remote middleware.

**Review by sfunkhouser — APPROVED** [review]
LGTM

File: ginauth/remote.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	unmarshallerr := json.Unmarshal(body, &authResp)
+	if unmarshallerr != nil {
+		return ClaimMetadata{}, NewAuthenticationError(unmarshallerr.Error())
+	}
+
+	if !authResp.Authed {
+		return ClaimMetadata{}, NewAuthenticationError(authResp.Message)
+	}
+
+	// TODO(jaosorior): Should we fail the request if no appropriate
```
[the remark this answers is not in the document]
Comment (JAORMX) [reply ginauth/remote.go:110]: that's right

**Pull request merged**